torch/csrc/autograd/engine.cpp

```diff
@@ -729,15 +649,30 @@ auto Engine::execute(const edge_list& roots,
   ClearCallbacks _cb_guard(final_callbacks_, post_callbacks_lock_);
 
   GraphTask graph_task(keep_graph, create_graph, worker_device == NO_DEVICE ? 0 : total_depth+1);
-  // Lock mutex while GraphTask is being set up
-  std::unique_lock<std::mutex> lock(graph_task.mutex_);
 
   // Now compute the dependencies for all executable functions and queue the root
   auto graph_root = std::make_shared<GraphRoot>(roots, inputs);
   compute_dependencies(graph_root.get(), graph_task);
+
   if (!outputs.empty()) {
     graph_task.init_to_execute(*graph_root, outputs);
   }
+  return execute_with_graph_task(graph_task, graph_root);
+}
+
+void Engine::enqueue_on_cpu(NodeTask task) {
+  std::call_once(start_threads_flag_, &Engine::start_threads, this);
+  ready_queue(at::kCPU).push(
+      std::move(task), /* incrementOutStandingTasks */ false);
+}
+
+variable_list Engine::execute_with_graph_task(
+    GraphTask& graph_task,
+    std::shared_ptr<Node> graph_root) {
+  std::call_once(start_threads_flag_, &Engine::start_threads, this);
```

Comment:
If you're going to stick thread initialization here, maybe stop initializing in `Engine::execute`?